Tilt Effect as suggested in #17

ShankarBUS opened the pull request:
This PR is for adding the **Pointer Up and Down Theme Animations**

1.  Preview (ModernWpf)
- Normal
 ![image](https://user-images.githubusercontent.com/54765038/75870256-c385fa00-5e30-11ea-84f8-a4238ee7bdd4.png)
- Pressed
 ![image](https://user-images.githubusercontent.com/54765038/75865259-4acf6f80-5e29-11ea-8fe1-110fd5fe7bd6.png)
2. Preview (UWP)
- Pressed
![image](https://user-images.githubusercontent.com/54765038/75870657-4c049a80-5e31-11ea-9f62-4971ed983be9.png)

> Note : for using this effect, the element to be affected must be placed with in a **Panel** (such as Grid, Canvas, etc.) not other controls. The effect is subtle

These animations can be implemented in Visual States as follows 

```xaml
    <VisualStateManager.VisualStateGroups>
        <VisualStateGroup x:Name="CommonStates">
            <VisualState x:Name="Normal">
                <Storyboard>
                    <animation:PointerUpThemeAnimation TargetName="ElementToBeAffected"/>
                </Storyboard>
            </VisualState>
            <VisualState x:Name="MouseOver">
                <Storyboard>
                    <animation:PointerUpThemeAnimation TargetName="ElementToBeAffected"/>
                </Storyboard>
            </VisualState>
            <VisualState x:Name="Pressed">
                <Storyboard>
                    <animation:PointerDownThemeAnimation TargetName="ElementToBeAffected"/>
                </Storyboard>
            </VisualState>
            <VisualState x:Name="Disabled">
            </VisualState>
        </VisualStateGroup>
    </VisualStateManager.VisualStateGroups>
```
where xmlns:animation = "ModernWpf.Media.Animation"

The contents may be blurred since they are placed in a ViewPort3D but not as blur as in UWP

Thanks to @ORRNY66 for suggesting this 😁.

![7d3dd84-1](https://user-images.githubusercontent.com/54765038/75866699-73f0ff80-5e2b-11ea-9b76-05ed00b08c3b.jpg)

ShankarBUS commented:
Since someone proposed to remove this effect from WinUI repo a year ago #microsoft/microsoft-ui-xaml#242, I don't know whether its a good choice to add this. But this is the only Library which brings the UWP UI to WPF. That's why I want you to implement this effect (Pleaseeeeeeee 🥺🙏)

ShankarBUS commented on ModernWpf/Styles/Button.xaml:
```diff
@@ -27,26 +28,49 @@
             <Setter.Value>
                 <ControlTemplate TargetType="Button">
-                    <Border
+                    <Grid>
```
Important Note : The PlaneratorHelper places the Planerator within a **Panel** control (It's designed that way sorry!) so be careful when using this effect

ShankarBUS commented on ModernWpf/Media/Animation/PointerUpThemeAnimation.cs:
```diff
@@ -0,0 +1,52 @@
+namespace ModernWpf.Media.Animation
+{
+    /// <summary>
+    /// Represents the preconfigured pointer up animation that applies to controls when
```
Typo error : have to change 'when' to 'after' and 'pressed' to 'released' in the xml comment

ShankarBUS commented on ModernWpf/Media/Animation/PointerUpThemeAnimation.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    /// <summary>
+    /// Represents the preconfigured pointer up animation that applies to controls when
+    /// they are pressed.
```
Typo error : have to change 'when they are pressed' to 'after they are released'

Kinnara commented on ModernWpf/Controls/Primitives/PointerUpDownHelper.cs:
```diff
@@ -0,0 +1,90 @@
+
+namespace ModernWpf.Controls.Primitives
+{
+    public static class PointerUpDownHelper
```
Could this class be made internal?

ShankarBUS replied:
Do you also want the **Planerator** and **PlaneratorHelper** to be `internal` classes? Because they are unnecessary to this library since you are trying to replicate **WinUI**

ShankarBUS replied:
If you elaborate your request, I will make those changes within a day

Kinnara commented:
Thanks a lot for this addition. I've seen that proposal as well and personally I'm with this comment: https://github.com/microsoft/microsoft-ui-xaml/issues/242#issuecomment-458822817. What about making the tilt effect opt-in? I can add a sample page for it once this is merged.

ShankarBUS commented:
Can you please elaborate your request?

From what I understood :

- You want the **PointerUpDownHelper** to be a `internal `Class
- You want the effect to be optionally turned on

The effect can be disabled by introducing a new attached dependency property **IsEnabled** (which will be _false_ by default), thus this effect can be enabled on individual elements :)

If PointerUpDownHelper is made internal we won't be able to disable/enable it from other projects

IMO Let's make the **IsPressed** Property `internal` so that won't cause any bugs and the effect can be enabled on an element or a style

ShankarBUS commented:
**Unnecessary Things to Know :**
Hey @Kinnara, **(others can skip this comment)**
You can understand why I used Pointer **animations** instead of manually handling it by checking this project [WPF Tilt Effect.zip](https://github.com/Kinnara/ModernWpf/files/4333800/WPF.Tilt.Effect.zip) (this is my modified version of the original code : [ZAxis.zip](https://github.com/Kinnara/ModernWpf/files/4333801/ZAxis.zip)). The original one had a **TiltEffect** [**Behaviour**](https://docs.microsoft.com/en-us/previous-versions/visualstudio/design-tools/expression-studio-4/ff726530(v=expression.40)) which is hard to remove from user side so I added a **TiltHelper** which can easily controlled (can apply tilt effect to any controls - borders, grid, textblock, textboxes and etc.). But the **TiltHelper** (found in my code) handled the mouse up/down/move events of the target element unlike the TiltEffect Behaviour (found in the original code) which process things in [**CompositionTarget.Rendering Event**](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.compositiontarget.rendering?view=netframework-4.8) (thus no need to handle mouse events). And Since Buttons can be also pressed by Keys (space, enter, or any...) it will be bad to use the TiltHelper. That's why I used **PointerUpDownHelper** and **animations** (just **boolean animations**) to handle the effect. The pressed state is triggered by the button itself and not by manual handling. Thus this is the efficient way.

 It may be hard for you to implement this into all button-like controls (adding visual states and animations) but the users can toggle this effect easily (users don't need to **re-template** the whole style as in **UWP/WinUI** just to remove the animations instead they can make the animations ineffective by setting **ui:TiltEffect.IsEnabled = "false"**)

**Finally, for achieving this effect :** The implementation is the same as in **WinUI** (only the TiltEffect.IsEnabled is additional for the users to toggle the effect, it only affect the pointer animation which are existing in the control template)
Check the Button style in this PR, the border "**Background**" has to be in a **Grid** (the button can be in any control, no restriction), the animations has to be in respective Visual states, the border "**Background**" must have this property `primitives:TiltEffect.IsEnabled="{TemplateBinding primitives:TiltEffect.IsEnabled}"` and the style must have `<Setter Property="primitives:TiltEffect.IsEnabled" Value="True"/>` (true or false it depends on your need). If it is done this way user can control it from individual controls or from a style. Users will be able to create custom styles based on default styles with/without tilt (_no re-templating_) like this
```xaml
<Style x:Key="MyButtonStyle" TargetType="Button" BasedOn="{StaticResource DefaultButtonStyle}">
        <Setter Property="primitives:TiltEffect.IsEnabled" Value="True"/>
</Style>
```
You can also apply this effect to **AppBarButton, AppBarToggleButton, DropDownButton, RepeatButton, ToggleButton** (also to **MenuItem, ListItem, ComboBoxItem** if you can) or to custom controls with Pressed Visual state (it is not a button-only effect, you can even animate a TextBlock if it has the pressed visual state).

In this PR I implemented this only to **Button**, I left the rest of the button-based controls for you to decide whether the others should have it or not

andrew-boyarshin commented:
Why is this closed? It's a good feature.

Kinnara commented:
I’m puzzled as well. I proposed some changes that would eliminate the need to re-template, and they could be implemented by me to save the OP some work. Shortly after that this PR was closed.

mikecoxlong commented:
All that just for a subtle effect. I wouldn't recommend adding this. Since the 3D parent control is added and removed whenever the animation is played,  it would cause some memory leakage and lag. No would even notice it 🤦🏻‍♂️

andrew-boyarshin commented:
But it would be great to have this at least opt-in. The existence of choice is always* better than it's absence.

ShankarBUS commented:
Looks like the **Tilt Effects** are finally gone from WinUI 🙄 https://github.com/microsoft/microsoft-ui-xaml/pull/3514.
